icu4c/source/common/uloc.cpp

```diff
@@ -1959,7 +1918,34 @@ uloc_getBaseName(const char* localeID,
                  int32_t nameCapacity,
                  UErrorCode* err)
 {
-    return _canonicalize(localeID, name, nameCapacity, _ULOC_STRIP_KEYWORDS, err);
+    if (U_FAILURE(*err)) {
+        return 0;
+    }
+
+    CheckedArrayByteSink sink(name, nameCapacity);
+    ulocimp_getBaseName(localeID, sink, err);
+
+    int32_t reslen = sink.NumberOfBytesAppended();
+
+    if (U_FAILURE(*err)) {
+        return reslen;
+    }
+
+    if (sink.Overflowed()) {
+        *err = U_BUFFER_OVERFLOW_ERROR;
+    } else {
+        u_terminateChars(name, nameCapacity, reslen, err);
```

Comment:
Optional: Might be useful to add a bytesinkutil function to terminateByteSink. Maybe specific to a CheckedArrayByteSink, subsuming all of the stuff like you have here lines 1928..1940 (but `sink.Append('\0')` for the actual termination). Then you could just `return ByteSinkUtil::terminateCheckedArrayByteSink(sink, *err);`

Reply:
Something along those lines could indeed be useful. I do however think that it for the time being is a good idea to simply continue copy-pasting this recipe until all the locale memory management has been updated, for at that point in time we'll have a much clearer view of exactly what would be the most useful design for such a helper function.

Reply:
> I do however think that it for the time being is a good idea to simply continue copy-pasting this recipe until all the locale memory management has been updated, for at that point in time we'll have a much clearer view of exactly what would be the most useful design for such a helper function.

PR #2863